fpnge.cc

```diff
@@ -14,13 +14,50 @@
 
 #include <algorithm>
 #include <assert.h>
-#include <immintrin.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <vector>
 
+
+#ifdef _MSC_VER
+# define FORCE_INLINE [[msvc::forceinline]]
+static FORCE_INLINE unsigned BSF32(unsigned v) {
+  unsigned long idx;
+  _BitScanReverse(&idx, v);
+  return idx;
+}
+# define __SSE4_1__ 1
+#else
+# define FORCE_INLINE __attribute__((always_inline))
+# define BSF32 __builtin_ctzl
+#endif
+
+
+#include <wmmintrin.h>  // for CLMUL
+
+#ifdef __AVX2__
+# include <immintrin.h>
+# define _mm(f) _mm256_##f
```

Comment:
Thanks for the feedback. Do you have an exact suggestion of what you'd prefer there?  
(also feel free to make any edits yourself, if you prefer)

The change was mostly meant to be a relatively unobtrusive way to introduce SSE4 support, so I didn't want to introduce further abstraction layers. (I'm personally not a fan of such anyway, as I prefer writing separate implementations for different ISAs, but you may prefer otherwise)

Reply:
After further thinking, an abstraction layer is probably unwarranted here, so I'd go with just MM-style macros until either me, you or someone else wants to extend this to other SIMD instruction sets :)

Reply:
So, rename `_mm` to `MM`, `_mmsi` to `MMSI` and `__mivec` to `MIVEC` - sound good to you?

Reply:
Yup!